okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java

```diff
@@ -170,32 +172,44 @@ private RealConnection findConnection(int connectTimeout, int readTimeout, int w
         refusedStreamCount = 0;
       }
     }
-    RealConnection newConnection = new RealConnection(connectionPool, selectedRoute);
+
+    RealConnection result = new RealConnection(connectionPool, selectedRoute);
 
     synchronized (connectionPool) {
-      acquire(newConnection);
-      Internal.instance.put(connectionPool, newConnection);
-      this.connection = newConnection;
+      acquire(result);
+      Internal.instance.put(connectionPool, result);
```

Comment:
Sorry, it may be more trouble than it's worth. Here were my notes.

Test program: https://gist.github.com/dave-r12/d029c81e927b56a23bcc4b26a9f53b4c

Given connections (C1, C2, C3):
- Break points on StreamAllocation: 121, 176, 184.
- Let C1 proceed.
- Let C2 proceed.
- Let C2 proceed (now in the pool but stalled on connect).
- Let C1 proceed.
- Let C1 proceed (C1 is now the primary H2 connection).
- Step over C2 (now connected, allocationLimit is upped for H2).
- Step over C3 (it found the connection in the pool from C2.)
- Let C2 proceed.